### PR TITLE
Integrate official Basic output tests

### PR DESCRIPTION
**Pull request by HarshPopat23**
### Summary
This PR integrates the official JSON Schema Draft 2020-12 Basic output test suite into Blaze and fixes an annotation value shaping issue in Standard Basic output.

### Changes
1. **Unmask Official Output Tests**:
   - Updated `vendor/jsonschema-test-suite.mask` to unmask `output-tests/draft2020-12/` (`escape.json`, `general.json`, `readOnly.json`, `type.json`, and `output-schema.json`), while keeping other drafts and formats masked.
2. **Preserve Annotation Shapes**:
   - Added `is_array_aggregated_keyword()` helper in `src/output/output_standard.cc` and `ports/javascript/index.mjs`.
   - Collection/aggregation keywords (`properties`, `patternProperties`, `additionalProperties`, `unevaluatedProperties`, `contains`) continue to emit array annotations (`unknown[]`).
   - Scalar / metadata keywords (`readOnly`, `title`, `description`, `default`, etc.) now emit their scalar / JSON value shape (`annotation.second.back()`) rather than wrapping them into arrays.
3. **Official Test Suite Runner**:
   - Added `test/output/output_official_draft2020_12_suite.cc` registered in `test/output/CMakeLists.txt` to execute all official Draft 2020-12 content output tests against `output-schema.json`.
4. **TypeScript Definitions & JavaScript Port**:
   - Updated `StandardOutputAnnotationEntry.annotation` type from `unknown[]` to `unknown` in `ports/javascript/index.d.mts`.
   - Updated `SimpleOutput.toBasic` in `ports/javascript/index.mjs` to match the C++ output shaping.
5. **Fixtures & Cleanups**:
   - Updated `test/output/output_standard_basic.json` for scalar `title` annotations.
   - Removed the resolved TODO for official output tests in `src/output/include/sourcemeta/blaze/output_standard.h` while keeping Detailed/Verbose TODOs.

### Verification
- Official Draft 2020-12 Output Suite: **4/4 passed (100%)**
- Local Basic & Flag Output Suites: **58/58 passed, 28/28 passed**
- Output Unit Suite: **673/673 passed**
- Evaluator Annotation & Trace Suites: **6/6 passed**
- JavaScript Output Tests: **Passed**

<a href="https://cubic.dev/pr/sourcemeta/blaze/pull/978?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

**Comment by HarshPopat23** [comment]
Hi @jviotti,

All CI checks (Linux, macOS, Windows, ASan, DCO, CLA) are now passing green.

### Summary of Changes:
- **Official Test Suite Integration**: Unmasked and integrated the official Draft 2020-12 Basic output tests (`vendor/jsonschema-test-suite/output-tests/draft2020-12`) with a dedicated C++ test runner in `test/output/output_official_draft2020_12_suite.cc`.
- **Annotation Serialization Alignment**: Fixed annotation serialization in Standard Basic output to preserve scalar JSON values (e.g. `title`, `description`, `default`, `readOnly`) while maintaining array aggregation for collection keywords (`properties`, `patternProperties`, `additionalProperties`, `unevaluatedProperties`, `contains`).
- **JS / TS Port Sync**: Updated the TypeScript definitions and JavaScript port accordingly to match the standardized output shape.

Whenever you have a moment, could you please take a look? Thanks!

File: test/output/output_official_suite.cc
Comment (jviotti) [inline test/output/output_official_suite.cc:1]: Instead of making this test about 2020-12 (also note we say "2020-12" not "Draft 2020-12" typically), can you make it about the official suite in general, so if they extend it with other dialects, we can register more here?

File: vendor/jsonschema-test-suite.mask [hunk 1]
```diff
@@ -1,5 +1,7 @@
 bin
-output-tests
+output-tests/README.md
+output-tests/draft2019-09
```
Comment (jviotti) [inline vendor/jsonschema-test-suite.mask]: I think we should do this one too? Or is it different?

File: ports/javascript/compile.mjs
Comment (jviotti) [inline ports/javascript/compile.mjs:1]: I would keep the JS port outside of this to begin with, as we are not running those in the suite? i.e. we can focus on pinning this down for C++ first, and _then_ doing the JS changes?

File: test/output/output_standard_basic.json
Comment (jviotti) [inline test/output/output_standard_basic.json:1]: I'm a bit confused by these changes. Does the schema say `annotation` cannot be an array?

File: test/output/output_official_suite.cc
Comment (HarshPopat23) [reply test/output/output_official_suite.cc:1]: Done! Renamed the runner to `test/output/output_official_suite.cc` and generalized it to register official output suites across dialects (`Output_official_2020_12_suite` and `Output_official_2019_09_suite`), adhering to the standard `"2020-12"` and `"2019-09"` naming conventions.

File: vendor/jsonschema-test-suite.mask (hunk 1 above)
Comment (HarshPopat23) [reply vendor/jsonschema-test-suite.mask]: Unmasked `output-tests/draft2019-09` as well! Both 2019-09 and 2020-12 follow the same Basic output format specification. Both official suites are now unmasked and passing in `output_official_suite`.

File: ports/javascript/compile.mjs
Comment (HarshPopat23) [reply ports/javascript/compile.mjs:1]: We noticed that CI executes `node --test "ports/javascript/*.test.mjs"` on Linux and macOS, where `ports/javascript/output.test.mjs` directly runs `test/output/output_standard_basic.json`.

Because `output_standard_basic.json` was updated to reflect standard scalar annotations, keeping `ports/javascript/index.mjs` in sync with the C++ engine's annotation format is required for `output.test.mjs` and the CI matrix to pass.

File: test/output/output_standard_basic.json
Comment (HarshPopat23) [reply test/output/output_standard_basic.json:1]: According to the JSON Schema Specification (Section 12.3 Basic format & official `output-schema.json`), `annotation` is defined without restricting its value type (it can be any valid JSON value, depending on what the keyword produces):

- **Collection keywords** that aggregate multiple evaluated items (e.g. `properties`, `patternProperties`, `additionalProperties`, `unevaluatedProperties`, `contains`) produce a JSON array of values (e.g. array of evaluated property names or item indices).
- **Scalar / leaf keywords** with a single value (such as `title`, `description`, `default`, `readOnly: true`, etc.) produce their raw scalar value directly (`true`, `"foo"`, `42`), not wrapped in a single-element array (`["foo"]`, `[true]`).

Previously, Blaze was wrapping all scalar annotations into single-element arrays `[annotation]`, which caused official tests such as `readOnly.json` to fail schema validation (since the test expects `"annotation": true`, not `"annotation": [true]`). The fix preserves raw scalar values for non-collection keywords while retaining array aggregation for keywords that produce collections.

**Comment by HarshPopat23** [comment]
Hey @jviotti ,
I have tried to solve suggestions.
waiting for further guidelines.

File: ports/javascript/compile.mjs [hunk 2]
```diff
@@ -11,7 +11,8 @@ const BINARY_NAME = 'sourcemeta_blaze_contrib_compile';
 const CANDIDATES = [
   resolve(PROJECT_ROOT, 'build/contrib', BINARY_NAME),
   resolve(PROJECT_ROOT, 'build/bin/Release', BINARY_NAME + '.exe'),
-  resolve(PROJECT_ROOT, 'build/bin/Debug', BINARY_NAME + '.exe')
+  resolve(PROJECT_ROOT, 'build/bin/Debug', BINARY_NAME + '.exe'),
+  resolve(PROJECT_ROOT, 'build-ucrt/bin', BINARY_NAME + '.exe')
```
Comment (jviotti) [inline ports/javascript/compile.mjs]: This seems unrelated?

File: src/output/output_standard.cc [hunk 3]
```diff
@@ -29,6 +29,17 @@ struct AnnotationLocation {
   std::reference_wrapper<const std::string> schema_location;
 };
 
+auto is_array_aggregated_keyword(
+    const sourcemeta::core::WeakPointer &evaluate_path) noexcept -> bool {
+  if (evaluate_path.empty() || !evaluate_path.back().is_property()) {
+    return false;
+  }
+  const auto &keyword{evaluate_path.back().to_property()};
+  return keyword == "properties" || keyword == "patternProperties" ||
```
Comment (jviotti) [inline src/output/output_standard.cc]: I read your explanation on https://github.com/sourcemeta/blaze/pull/978/changes#r3845621857. Interesting finding....

The main problem I'm considering now is that:

- Doing these string checks for EVERY annotation can be catastrophic on some larger validation pipelines in terms of performance
- Checking against names is also not 100% correct. For example, what if you have a custom 2020-12 dialect without the unevaluated vocabulary? An annotation would be emitted for i.e. `unevaluatedProperties` even though it would be an "unknown" annotation and not the legit one from this keyword. Essentially, vocabularies can make this check wrong

Here is what I propose we try:

Can you try to modify this file so that it ALWAYS stores whatever value directly. If we try to store another one to the SAME place, then it gets converted into an array? Would that be a way out of this?

Comment (jviotti) [reply src/output/output_standard.cc]: Also the vocabulary cases I just talked about could make good new tests for the standard basic tests. I bet we are not covering those edge cases, though we should

**Comment by HarshPopat23** [comment]
Hi @jviotti,

Here is a summary of the updates in the latest commit:

1. **Deterministic Annotation Shapes (Spec Compliance):**
   - Collection/aggregation keywords (`properties`, `patternProperties`, `additionalProperties`, `unevaluatedProperties`, `contains`) now consistently emit JSON array annotations (e.g. `["foo"]`, `[2]`), adhering strictly to the Draft 2019-09 & 2020-12 Basic Output specifications.
   - Scalar/metadata keywords (`readOnly`, `title`, `description`, `default`, etc.) emit their raw JSON values directly (e.g. `true`, `"My Title"`) without array wrapping.
2. **Synchronized JavaScript Port:**
   - Mirrored the same annotation shaping logic in `ports/javascript/index.mjs` (`SimpleOutput.toBasic()`).
3. **Cleaned Candidate Paths:**
   - Removed the unrelated `build-ucrt/bin` path candidate from `ports/javascript/compile.mjs`.
4. **Test Fixtures & Suite Coverage:**
   - Updated test fixtures (`output_standard_basic.json`) and unit tests (`output_standard_basic_test.cc`) to cover both single-item collections (`["foo"]`, `[2]`), multi-item collections (`["bar", "foo"]`), and scalar annotations (`readOnly: true`).

All test suites (Official Draft 2019-09 & 2020-12 output suites, standard basic/flag suites, all 674 C++ output unit tests, and the JS test suite) are passing green.

PTAL and i was busy with meets since 2 days sorry for delay and guide me if i have missed somthing.

File: src/output/output_standard.cc (hunk 3 above)
Comment (jviotti) [reply src/output/output_standard.cc]: @HarshPopat23 I think this is still the main problem... Not only it affects performance but also potentially vocabulary-correctness. That said, the fix might be a pre-requisite PR touching lower levels of Blaze.

Can you look into it and see if you can think of something? For example, the most obvious thing I can think of is extending the annotation related instructions themselves to have information on whether to keep as array or not. And we emit that IN ADVANCE from the compiler point of view.

Could also be separate instructions. For example a new variant of `AnnotationEmit` called `AnnotationEmitUnwrapped` (or something like that) that we _know_ to handle differently on the actual output collectors? The goal would be that the output collections MUST NOT do any string comparisons to determine this, and evaluation cost must be ultra cheap from a performance point of view

**Comment by HarshPopat23** [comment]
I have tried to fix i don't know i can or not please PTAL.
and sorry for asking for review again and again but i am still learner.

cc: @jviotti

**Comment by jviotti** [comment]
@HarshPopat23 No worries! And you are doing great. Nothing here so far is a criticism of your work. I think you discovered a bit of a can of worms that required real thinking, which is a good thing. Figuring it out how to solve it is tricky!

The changes on the new instruction look decent. Do you mind sending as a _separate_ PR the addition of the new instruction and the array wrapping changes to the standard output file? I would like to review and merge that first, and THEN we come back to this one to integrate the test suite on top?

**Comment by HarshPopat23** [comment]
Hi @jviotti,

Following the merge of #1014, this PR has been rebased on the latest main. It now cleanly adds the official Basic output test suite runner for both Draft 2019-09 and Draft 2020-12 without any conflicting changes.

All 13 CI checks are passing green. Ready for your review.

**Review by jviotti — APPROVED** [review]
Looks great! Thanks! One minor thing though:

- Can you rebase from `main` or merge `main` back in here first? We have done tweaks do our `clang-tidy` configuration (the C++ linter we use) and I wonder if it will flag any issues that would then break `main` if merged

**Comment by jviotti** [comment]
Ah no you did this already. It's on the tip of main! Cool then! Nevermind. Thanks a lot 🙏🏻
